src/compiler/visitorPublic.ts

```diff
@@ -189,19 +267,19 @@ function visitArrayWorker<T extends Node>(nodes: readonly T[], visitor: Visitor,
             if (visited) {
                 if (isArray(visited)) {
                     for (const visitedNode of visited) {
-                        void Debug.assertNode(visitedNode, test);
-                        updated.push(visitedNode as T);
+                        updated.push(visitedNode);
                     }
                 }
                 else {
-                    void Debug.assertNode(visited, test);
-                    updated.push(visited as T);
+                    updated.push(visited);
                 }
             }
         }
     }
 
-    return updated ?? nodes;
+    const result = updated ?? nodes;
+    Debug.assertEachNode(result, test);

```

Comment:
Asserting late rather than early means that a lot of work could happen before we report an error.

Reply:
Yeah, this was an open question I had in an earlier thread here (it might still be open). I will change this back and just cast things to make it work.

Reply:
I remembered why I left this; if we don't do it at the end, we won't be able to tell if the output array consists of nodes which match the test if the visitor didn't change any of the nodes. Do we want to check early on each modify instead and just keep the second pass?

Reply:
I've modified this to restore the eager checks. The only missing checks I am now adding are when `updated` is set for the first time as a slice of the original array, or when the original array is about to be returned instead. Those are the only cases where we weren't previously testing things.